play-akka-http-ws-standalone/src/main/java/play/libs/ws/akkahttp/StandaloneAkkaHttpWSRequest.java

```diff
@@ -0,0 +1,585 @@
+/*
+ * Copyright (C) 2009-2017 Lightbend Inc. <https://www.lightbend.com>
+ */
+package play.libs.ws.akkahttp;
+
+import akka.actor.ActorSystem;
+import akka.http.impl.model.parser.HeaderParser$;
+import akka.http.javadsl.Http;
+import akka.http.javadsl.model.*;
+import akka.http.javadsl.model.headers.BasicHttpCredentials;
+import akka.http.scaladsl.model.HttpHeader;
+import akka.http.scaladsl.model.HttpHeader$;
+import akka.pattern.PatternsCS;
+import akka.stream.Materializer;
+import play.libs.ws.*;
+import scala.concurrent.duration.FiniteDuration;
+
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+public final class StandaloneAkkaHttpWSRequest implements StandaloneWSRequest {
```

Comment:
I think we talked about that already, but is there any good reason to *implement* these classes in Java? Keeping the interfaces written in Java might make sense but the implementation could still be done in Scala, right? If that's not possible I would still prefer an approach where the Java classes just delegate to the real implementation which could then be written in Scala.

Reply:
This is not possible that easily:

This was an expirment i did in the past: https://github.com/schmitch/play-ws-akka
The biggest probem is that if something returns `java.lang.Boolean` instead of `boolean` or everything that returns a `Future`/`CompletionStage` I created a `WrappedFuture` to support that: https://github.com/schmitch/play-ws-akka/blob/master/play-ws-akka-standalone/src/main/scala/play/api/libs/ws/akkaws/WrappedFuture.scala

The delegate approach might be more feasible.

Reply:
At least doing it this way uncovered some missing Java API methods in Akka Http.

Reply:
>  is there any good reason to implement these classes in Java

Similar to what @2m said, the main reason is idiom -- if you're changing or adding to the API in Java but writing the tests in Scala, you often don't feel the pain of refactoring because you can leverage the Scala Collections API.  It also helps for discovery, because Java users don't have to decipher Scala tests to see how the code should work, and writing documentation, because the tests and examples match up.

Reply:
I was only talking about implementation. Interfaces and tests could and probably should be done in Java, I agree.